src/org/nejrasm/zadaca3oop/task2/ElectricCar.java

```diff
@@ -0,0 +1,33 @@
+package org.nejrasm.zadaca3oop.task2;
+
+public class ElectricCar extends Vehicle {
+    private final ElectricEngine electricCarEngine;
+
+    public ElectricCar(final String model, final double capacity) {
+        super(model);
+        this.electricCarEngine = new ElectricEngine(capacity);
+    }
+
+    @Override
+    public double accelerate(final double acceleration) {
+        final double capacityState = this.electricCarEngine.decreaseCapacity(0.4 * acceleration / 100);
+        if ((this.electricCarEngine.capacity - capacityState) > 0) {
+            this.electricCarEngine.capacity = capacityState;
+            super.accelerate(acceleration);
+        }
+        return this.speed;
+    }
+
+    @Override
+    public double decelerate(final double deceleration) {
+        super.decelerate(deceleration);
+        this.electricCarEngine.brake(deceleration / 100); //I have added here /100 to make more sens in calculations. Sample with increase is commented below.
+        //this.electricCarEngine.capacity = this.electricCarEngine.increaseCapacity(0.1*deceleration/100);
+        return this.speed;
```

Comment:
Small bug here, the `decelerate()` method will not necessary decrease the speed by `deceleration`.

If this.speed < deceleration, we only decelerate by `deceleration - speed`, meaning the brake method will charge the car less than you calculated.

You might want to refactor, and return from accelerate and decelerate the value by which the car increased/decreased it's speed, rather than the current speed

Reply:
I see, I have changed the logic now. Furthermore, it deceleration (the amount we want to lower our speed for) is higher than the speed we have at the moment (this.speed) then we can only decelerate to zero, meaning that the final/max deceleration could be actually equal to this.speed (the value we could further use to calculate charge amount), and the final speed will be 0. Am I right or there is other logic here with this charging part?